### PR TITLE
[SMALLFIX] Preserve permissions with no recursive parent directory creation in UFS

### PR DESCRIPTION
**Pull request by madanadit**
In this PR:
- UFS.create has a new createParent option to safeguard against permissions issues with recursive parent directory creation. 
- Persist on worker in case of multiple mount points had a potential bug.

### CODE REVIEW

File: core/common/src/main/java/alluxio/underfs/options/CreateOptions.java [hunk 1]
```diff
@@ -41,11 +44,19 @@ public static CreateOptions defaults() {
    * Constructs a default {@link CreateOptions}.
    */
   private CreateOptions() {
+    mCreateParent = true;
```
Comment (ifcharming) [inline core/common/src/main/java/alluxio/underfs/options/CreateOptions.java]: is it possible to make `mCreateParent` default to false?
Seems like in most cases for LocalFS and HDFS, we explicitly ensure the caller sites to set this option to `false`. Would setting the default to false help identify any other places where this is set to true by accident?

File: core/server/src/main/java/alluxio/worker/file/FileDataManager.java [hunk 2]
```diff
@@ -302,42 +301,13 @@ public void persistFile(long fileId, List<Long> blockIds) throws AlluxioExceptio
...
-        // and assume the directory is already prepared, regardless of permission matching.
-        if (!mUfs.mkdirs(ufsDirAndPerm.getFirst(), ufsDirAndPerm.getSecond())
-            && !mUfs.isDirectory(ufsDirAndPerm.getFirst())) {
-          throw new IOException("Failed to create dir: " + ufsDirAndPerm.getFirst());
-        }
-      }
-    }
-    return dstPath.toString();
+    AlluxioURI alluxioPath = new AlluxioURI(fileInfo.getPath());
+    FileSystem fs = FileSystem.Factory.get();
```
Comment (ifcharming) [inline core/server/src/main/java/alluxio/worker/file/FileDataManager.java]: thanks for fixing this -- we should not have assumed `UNDERFS_ADDRESS` is always the correct ufs prefix.

**Comment by madanadit** [comment]
@ifcharming I verified that prepareUFSFilePath is not required in FileOutStream as BaseFileSystem.createFile calls master.createFile before attempting to write the data file. The master persists any directories, the worker then can safely create just the file with ufs createoptions with createparent = false.

File: core/common/src/main/java/alluxio/underfs/options/CreateOptions.java [hunk 3]
```diff
@@ -24,6 +24,9 @@
 @PublicApi
 @NotThreadSafe
 public final class CreateOptions {
+  // Determine whether to create any necessary but nonexistent parent directories.
```
Comment (ifcharming) [inline core/common/src/main/java/alluxio/underfs/options/CreateOptions.java]: can you please add a comment to state the guideline when a caller should set to this `true`?

File: core/client/src/main/java/alluxio/client/file/UnderFileSystemUtils.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+/**
+ * Utility functions for working with {@link alluxio.underfs.UnderFileSystem}.
+ *
+ */
+@ThreadSafe
+public final class UnderFileSystemUtils {
+
+  /**
+   * Create parent directories for path with correct permissions if required.
```
Comment (ifcharming) [inline core/client/src/main/java/alluxio/client/file/UnderFileSystemUtils.java]: s/Create/Creates

File: core/client/src/test/java/alluxio/client/file/FileOutStreamTest.java [hunk 5]
```diff
@@ -279,9 +280,9 @@ public void cancelWithDelegation() throws Exception {
   @Test
   public void cancelWithoutDelegation() throws Exception {
     Configuration.set(PropertyKey.USER_UFS_DELEGATION_ENABLED, false);
-    OutStreamOptions options =
-        OutStreamOptions.defaults().setBlockSizeBytes(BLOCK_LENGTH)
-            .setWriteType(WriteType.CACHE_THROUGH).setPermission(Permission.defaults());
+    OutStreamOptions options = OutStreamOptions.defaults().setBlockSizeBytes(BLOCK_LENGTH)
```
Comment (ifcharming) [inline core/client/src/test/java/alluxio/client/file/FileOutStreamTest.java]: remove trailing space?

File: core/client/src/main/java/alluxio/client/file/FileOutStream.java [hunk 6]
```diff
@@ -141,8 +141,8 @@ public FileOutStream(AlluxioURI path, OutStreamOptions options, FileSystemContex
         } else {
           UnderFileSystem ufs = UnderFileSystem.Factory.get(mUfsPath);
           // TODO(jiri): Implement collection of temporary files left behind by dead clients.
-          CreateOptions createOptions =
-              CreateOptions.defaults().setPermission(options.getPermission());
+          CreateOptions createOptions = CreateOptions.defaults()
```
Comment (ifcharming) [inline core/client/src/main/java/alluxio/client/file/FileOutStream.java]: maybe revert this change? now it's only format change, which seems unnecessary.

File: core/client/src/main/java/alluxio/client/file/FileOutStream.java [hunk 7]
```diff
@@ -141,8 +141,8 @@ public FileOutStream(AlluxioURI path, OutStreamOptions options, FileSystemContex
         } else {
           UnderFileSystem ufs = UnderFileSystem.Factory.get(mUfsPath);
           // TODO(jiri): Implement collection of temporary files left behind by dead clients.
```
Comment (ifcharming) [inline core/client/src/main/java/alluxio/client/file/FileOutStream.java]: can you please add a comment here to explain why there's no need to call `prepareUfsPath` while `createParent` option is `false`?

Comment (madanadit) [inline core/client/src/main/java/alluxio/client/file/FileOutStream.java]: done

**Comment by madanadit** [comment]
@ifcharming Thanks for the review. Please take another look.

**Comment by madanadit** [comment]
@calvinjia Can you also take a look?. Thanks.

File: core/client/src/main/java/alluxio/client/file/UnderFileSystemUtils.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    AlluxioURI dstPath = new AlluxioURI(ufsPath);
+    String parentPath = dstPath.getParent().getPath();
+    // creates the parent folder if it does not exist
+    if (!ufs.isDirectory(parentPath)) {
+      // Create ancestor directories from top to the bottom. We cannot use recursive create parents
+      // here because the permission for the ancestors can be different.
+      Stack<Pair<String, MkdirsOptions>> ufsDirsToMakeWithOptions = new Stack<>();
+      AlluxioURI curAlluxioPath = alluxioPath.getParent();
+      AlluxioURI curUfsPath = dstPath.getParent();
+      // Stop at the Alluxio root because the mapped directory of Alluxio root in UFS may not exist.
```
Comment (calvinjia) [inline core/client/src/main/java/alluxio/client/file/UnderFileSystemUtils.java]: Should we stop at a mount point for the Alluxio path?

File: core/common/src/main/java/alluxio/underfs/BaseUnderFileSystem.java [hunk 9]
```diff
@@ -62,7 +62,7 @@ public void configureProperties() throws IOException {
 
   @Override
   public OutputStream create(String path) throws IOException {
-    return create(path, CreateOptions.defaults());
+    return create(path, CreateOptions.defaults().setCreateParent(true));
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/underfs/BaseUnderFileSystem.java]: Should `setCreateParent` be defaulted to true? It seems weird to have an implicit non-default option here.

File: tests/src/test/java/alluxio/underfs/UnderStorageSystemInterfaceIntegrationTest.java [hunk 10]
```diff
@@ -90,6 +91,22 @@ public void createEmpty() throws IOException {
...
+  public void createNoParent() throws IOException {
+    String testFile = PathUtils.concatPath(mUnderfsAddress, "testDir/testFile");
+    try {
+      OutputStream o = mUfs.create(testFile, CreateOptions.defaults().setCreateParent(false));
+      o.close();
+    } catch (IOException e) {
+      // IOException is expected
+      return;
+    }
+    Assert.fail("Create with no parent should throw an exception");
```
Comment (calvinjia) [inline tests/src/test/java/alluxio/underfs/UnderStorageSystemInterfaceIntegrationTest.java]: You can use the `ExpectedException` rule for convenience.

File: underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsUnderFileSystem.java [hunk 11]
```diff
@@ -154,7 +155,13 @@ public OutputStream createDirect(String path, CreateOptions options) throws IOEx
       try {
         LOG.debug("Creating HDFS file at {} with perm {}", path, perm.toString());
         // TODO(chaomin): support creating HDFS files with specified block size and replication.
-        return FileSystem.create(mFileSystem, new Path(path),
+        Path hdfsPath = new Path(path);
+        if (!options.getCreateParent()) {
```
Comment (calvinjia) [inline underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsUnderFileSystem.java]: Will the `FileSystem.create` call automatically create parent directories? Ideally we want to avoid the extra exists call

File: underfs/local/src/main/java/alluxio/underfs/local/LocalUnderFileSystem.java [hunk 12]
```diff
@@ -91,6 +92,12 @@ public OutputStream create(String path, CreateOptions options) throws IOExceptio
   @Override
   public OutputStream createDirect(String path, CreateOptions options) throws IOException {
     path = stripPath(path);
+    if (!options.getCreateParent()) {
+      File file = new File(path);
+      if (!file.getParentFile().exists()) {
```
Comment (calvinjia) [inline underfs/local/src/main/java/alluxio/underfs/local/LocalUnderFileSystem.java]: Similar question about the exists call

File: underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsUnderFileSystem.java (hunk 11 above)
Comment (madanadit) [inline underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsUnderFileSystem.java]: Yes, FileSystem.create will automatically create parent directories. HDFS has a call called createNonRecursive which is what we need. However, it was deprecated and then un-deprecated in 3.0.0-alpha1: http://hadoop.apache.org/docs/r3.0.0-alpha1/api/org/apache/hadoop/fs/FileSystem.html.

For HDFS vendors who have not implemented createNonRecursive, we could try non-recursive first and then fallback to the code above if an IOException is thrown. @calvinjia thoughts?.

File: core/client/src/main/java/alluxio/client/file/UnderFileSystemUtils.java (hunk 8 above)
Comment (madanadit) [inline core/client/src/main/java/alluxio/client/file/UnderFileSystemUtils.java]: True. The comment has gone stale.

If the mount point the current path is under is not deleted after mounting, the loop will stop at the mount point as ufs.isDirectory will return true. But you're right to handle the case where it might be, we need to stop at mount point.

File: core/common/src/main/java/alluxio/underfs/BaseUnderFileSystem.java (hunk 9 above)
Comment (madanadit) [inline core/common/src/main/java/alluxio/underfs/BaseUnderFileSystem.java]: I agree it is a little weird. @ifcharming and I did discuss this. For permissions, it makes sense to have a default = false. For journal and other test cases, which also use the same UFS API and don't care about permissions, it makes sense for a default = true. The latter case ends up using create w/o options. The former uses create w/ options as permissions are set.

Comment (calvinjia) [inline core/common/src/main/java/alluxio/underfs/BaseUnderFileSystem.java]: It seems like we would need to split out the journal interface from the ufs interface if we want to address this.

File: underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsUnderFileSystem.java [hunk 13]
```diff
@@ -154,7 +154,8 @@ public OutputStream createDirect(String path, CreateOptions options) throws IOEx
       try {
         LOG.debug("Creating HDFS file at {} with perm {}", path, perm.toString());
         // TODO(chaomin): support creating HDFS files with specified block size and replication.
-        return FileSystem.create(mFileSystem, new Path(path),
+        Path hdfsPath = new Path(path);
+        return FileSystem.create(mFileSystem, hdfsPath,
```
Comment (calvinjia) [inline underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsUnderFileSystem.java]: Should this be reverted?

File: core/client/src/main/java/alluxio/client/file/UnderFileSystemUtils.java [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+    AlluxioURI dstPath = new AlluxioURI(ufsPath);
+    String parentPath = dstPath.getParent().getPath();
+    // creates the parent folder if it does not exist
+    if (!ufs.isDirectory(parentPath)) {
+      // Create ancestor directories from top to the bottom. We cannot use recursive create parents
+      // here because the permission for the ancestors can be different.
+      Stack<Pair<String, MkdirsOptions>> ufsDirsToMakeWithOptions = new Stack<>();
+      AlluxioURI curAlluxioPath = alluxioPath.getParent();
+      AlluxioURI curUfsPath = dstPath.getParent();
+      // Stop at Alluxio mount point because the mapped directory in UFS may not exist.
```
Comment (calvinjia) [inline core/client/src/main/java/alluxio/client/file/UnderFileSystemUtils.java]: Could you add a unit test to valid the behavior in the possible situations?

**Comment by calvinjia** [comment]
@madanadit Looks good overall, could you increase the test coverage for the new method added?

File: core/client/src/test/java/alluxio/client/file/UnderFileSystemUtilsTest.java [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+import org.powermock.modules.junit4.PowerMockRunner;
+
+import java.io.IOException;
+
+/**
+* Tests {@link UnderFileSystemUtils}.
+*/
+@RunWith(PowerMockRunner.class)
+@PrepareForTest({FileSystem.class, UnderFileSystem.class})
+public class UnderFileSystemUtilsTest {
```
Comment (ifcharming) [inline core/client/src/test/java/alluxio/client/file/UnderFileSystemUtilsTest.java]: nit: final?

File: core/client/src/test/java/alluxio/client/file/UnderFileSystemUtilsTest.java [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+
+  private FileSystem mFileSystem;
+  private UnderFileSystem mUnderFileSystem;
+
+  // Test paths in Alluxio
+  AlluxioURI mAlluxioRoot;
```
Comment (ifcharming) [inline core/client/src/test/java/alluxio/client/file/UnderFileSystemUtilsTest.java]: nit: seems like all those URIs and UFS paths are static. What do you think of making these constant with full name? would the following format have slightly better readability?

```
final String mUfsRoot = "/ufs";
final String mUfsA = "/ufs/a";
final String mUfsB = "/ufs/a/b";
final String mUfsFile = "/ufs/a/b/file";
```

**Comment by ifcharming** [comment]
LGTM
